### DISCUSSION TITLE
Getting a server's response when the request status code is not 2XX

### DISCUSSION THREAD

**Discussion by ondomatej** [aio-libs/aiohttp#6558] [Q&A] [answered]
Hi, 

I'm using aiohttp to make async http requests, however I can't figure out how to get a response from the server when a request returns a 4XX error.

```py
    async def login(self, username: str, password: str) -> None:
        async with aiohttp.ClientSession(headers=self._headers) as session:
            self._session = session
            route = self._route + "/api/login"

            data = {
                        "grant_type": "password",
                        "client_id": "ANDR",
                        "username": username,
                        "password": password
                    }

            async with session.post(route, data=data, headers=self._headers) as resp:
                if resp.ok:
                    resp_json = await resp.json()
                    self._headers['Authorization'] = 'Bearer ' + json['resp_json']
                else:
                    resp_json = await resp.json()
                    raise InvalidGrant(response)
```

When the response returns a 2XX error (aka when `resp.ok` is `True`), then the code works just fine, but when the response is not OK (in this case: a 400 error), then `aiohttp.client_exceptions.ClientConnectionError: Connection closed` is raised. However, the server returns a specific error message when the request status code isn't 2XX, which is more descriptive than a generic `Bad Request` error. What should I do to get the server's response even when the status code isn't 2XX?

**Comment by adamko147** [comment, marked as answer]
I remember fixing this with https://github.com/aio-libs/aiohttp/pull/5407
does it work fine when you change your code as follows? 

```python
           async with session.post(route, data=data, headers=self._headers) as resp:
                resp_json = await resp.json()
                if resp.ok:
                    self._headers['Authorization'] = 'Bearer ' + json['resp_json']
                else:
                    raise InvalidGrant(response)
```

**Reply by ondomatej** [reply]
Yup, that solved it. I'm using aiohttp 3.7.4, which is why I'm still encountering this error.

Thanks!

**Reply by adamko147** [reply]
3.7 uses internally `raise_for_status` in `ok` property and therefore `self.release` when response status code > 400, so the connection is closed and you can't read the body anymore, see https://github.com/aio-libs/aiohttp/blob/3.7/aiohttp/client_reqrep.py#L999
